doc/designs/slh-dsa.md

```diff
@@ -0,0 +1,123 @@
+will be used to hold the keys (since the keys are only a maximum of 64 bytes each)
+
+struct slh_dsa_key_st {
+    /* The public key consists of a SEED and ROOT values each of size |n| */
+    uint8_t pub[SLH_DSA_MAX_KEYLEN];
+    /* The private key consists of a SEED and PRF values of size |n| */
+    uint8_t priv[SLH_DSA_MAX_KEYLEN];
+    size_t key_len; /* This value is set to 2 * n if there is a public key */
+    /* contains the algorithm name and constants such as |n| */
+    const SLH_DSA_PARAMS *params;
+    int has_priv; /* Set to 1 if there is a private key component */
+    ...
+};
+
+The fields `key_len` and `has_priv` are used to determine if a key has loaded
+the public and private key elements.
+The `params` field is the parameter set which is resolved via the algorithm name.
+
+In FIPS 205 the SLH_DSA private key contains the public key.
+In OpenSSL these components are stored separately, so there must always be a
+public key in order for the key to be valid.
+
+The key generation process creates a private key and half of the public key
+using DRBG's. The public key root component is then computed based on these
+values. For ACVP testing these values are supplied as an ENTROPY parameter.
+It is assumed that from data will not deal with a partial public key, and if this
+is required the user should use the key generation operation.
+
+Pure vs Pre Hashed Signature Generation
+----------------------------------------
+
+The normal signing process (called Pure SLH-DSA Signature Generation)
+encodes the message internally as 0x00 || len(ctx) || ctx || message.
+where `ctx` is some optional value of size 0x00..0xFF.
+
+ACVP Testing requires the ability for the message to not be encoded also. This
+will be controlled by settable parameters.
+
+Pre Hash SLH-DSA Signature Generation encode the message as
+
+```c
+0x01 || len(ctx) || ctx || digest_OID || H(message).
+```
+
+The scenario that is stated that this is useful for is when this encoded message
+is supplied from an external source.
+
+Currently I do not support the Pre Hash variant as this does not sit well with the
+OpenSSL API's. The user could do the encoding themselves and then set the settable
+to not encode the passed in message.
+
+Signing API
+-------------
+
+As only the one-shot implementation is required and the message is not digested
+the API's used should be
+
+EVP_PKEY_sign_message_init(), EVP_PKEY_sign(),
+EVP_PKEY_verify_message_init(), EVP_PKEY_verify().
+
+Buffers
+-------
+
+There are many functions pass buffers of size `n` Where n is one of 16,24,32
+depending on the algorithm name. These are used for key elements and hashes, so
+PACKETS are not used for these.
+
+Where it makes sense to, WPACKET is used for output (such as signature generation)
+and PACKET for reading signature data.
+
+Constant Time Considerations
+----------------------------
+
+As the security of SLH-DSA depends only on hash functions, I do not foresee
```

Comment:
I wouldn't use first person here - this is a consolidated view - not just yours - i.e. "we".

Reply:
"We" is still first person, just plural :grin: 
I agree with the comment though.